File: sdks/dotnet/Settled.Sdk.Tests/VerifierTests.cs
using System.Text.Json;
using Settled.Sdk;
using Xunit;

namespace Settled.Sdk.Tests;

public class VerifierTests
{
    private static readonly string VectorsDir = Path.GetFullPath(
        Path.Combine(AppContext.BaseDirectory, "../../../../../../test-vectors"));

    private static JsonElement Load(string name) =>
        JsonSerializer.Deserialize<JsonElement>(File.ReadAllText(Path.Combine(VectorsDir, name)));

    private static byte[] H(string hex)
    {
        var bytes = new byte[hex.Length / 2];
        for (var i = 0; i < hex.Length; i += 2)
            bytes[i / 2] = Convert.ToByte(hex.Substring(i, 2), 16);
        return bytes;
    }

    private static List<byte[]> Proofs(JsonElement arr) =>
        arr.EnumerateArray().Select(e => H(e.GetString()!)).ToList();

    // ── Leaf hashes ───────────────────────────────────────────────────────────

    [Fact]
    public void LeafHashVectors()
    {
        foreach (var v in Load("leaf-hashes.json").EnumerateArray())
        {
            var got = Verifier.LeafHash(H(v.GetProperty("input_hex").GetString()!));
            Assert.Equal(v.GetProperty("hash_hex").GetString(), ToHex(got));
        }
    }

    // ── Node hashes ───────────────────────────────────────────────────────────

    [Fact]
    public void NodeHashVectors()
    {
        foreach (var v in Load("node-hashes.json").EnumerateArray())
        {
            if (!v.TryGetProperty("hash_hex", out var expectedProp)) continue;
            var got = Verifier.NodeHash(
                H(v.GetProperty("left_hex").GetString()!),
                H(v.GetProperty("right_hex").GetString()!));
            Assert.Equal(expectedProp.GetString(), ToHex(got));
        }
    }

    [Fact]
    public void NodeHashNonCommutative()
    {
        foreach (var v in Load("node-hashes.json").EnumerateArray())
        {
            if (!v.TryGetProperty("swapped_hash_hex", out var swappedProp)) continue;
            var left = H(v.GetProperty("left_hex").GetString()!);
            var right = H(v.GetProperty("right_hex").GetString()!);
            var ab = Verifier.NodeHash(left, right);
            var ba = Verifier.NodeHash(right, left);
            Assert.NotEqual(ToHex(ab), ToHex(ba));
            Assert.Equal(swappedProp.GetString(), ToHex(ba));
        }
    }

    // ── Inclusion proofs ──────────────────────────────────────────────────────

    [Fact]
    public void InclusionProofVectors()
    {
        foreach (var v in Load("inclusion-proofs.json").EnumerateArray())
        {
            var ok = Verifier.VerifyInclusion(
                H(v.GetProperty("leaf_hash_hex").GetString()!),
                (ulong)v.GetProperty("leaf_index").GetInt64(),
                (ulong)v.GetProperty("tree_size").GetInt64(),
                Proofs(v.GetProperty("proof_hex")),
                H(v.GetProperty("root_hex").GetString()!));
            Assert.True(ok, $"size={v.GetProperty("tree_size")} idx={v.GetProperty("leaf_index")}");
        }
    }

    // ── Consistency proofs ────────────────────────────────────────────────────

    [Fact]
    public void ConsistencyProofVectors()
    {
        foreach (var v in Load("consistency-proofs.json").EnumerateArray())
        {
            var ok = Verifier.VerifyConsistency(
                (ulong)v.GetProperty("old_size").GetInt64(),
                (ulong)v.GetProperty("new_size").GetInt64(),
                Proofs(v.GetProperty("proof_hex")),
                H(v.GetProperty("old_root_hex").GetString()!),
                H(v.GetProperty("new_root_hex").GetString()!));
            Assert.True(ok, $"old={v.GetProperty("old_size")} new={v.GetProperty("new_size")}");
        }
    }

    // ── Signed Tree Heads ─────────────────────────────────────────────────────

    [Fact]
    public void SthVectors()
    {
        foreach (var v in Load("signed-tree-heads.json").EnumerateArray())
        {
            var treeSize = (ulong)v.GetProperty("tree_size").GetInt64();
            var rootHash = H(v.GetProperty("root_hash_hex").GetString()!);
            var ts = v.GetProperty("timestamp_ns").GetInt64();
            var sig = H(v.GetProperty("signature_hex").GetString()!);
            var pubKey = H(v.GetProperty("public_key_hex").GetString()!);
            Assert.True(Verifier.VerifyTreeHead(treeSize, rootHash, ts, sig, pubKey),
                v.GetProperty("description").GetString());
            Assert.False(Verifier.VerifyTreeHead(treeSize + 1, rootHash, ts, sig, pubKey),
                "tampered tree_size should fail");
        }
    }

    // ── Sequential STH verification ───────────────────────────────────────────

    [Fact]
    public void SthSequentialVectors()
    {
        var vectors = Load("signed-tree-heads.json").EnumerateArray().ToList();

        // Consecutive pairs must pass (timestamps are strictly increasing).
        for (var i = 0; i + 1 < vectors.Count; i++)
        {
            var prev = vectors[i];
            var curr = vectors[i + 1];
            var ok = Verifier.VerifyTreeHeadSequential(
                (ulong)curr.GetProperty("tree_size").GetInt64(),
                H(curr.GetProperty("root_hash_hex").GetString()!),
                curr.GetProperty("timestamp_ns").GetInt64(),
                H(curr.GetProperty("signature_hex").GetString()!),
                H(curr.GetProperty("public_key_hex").GetString()!),
                prev.GetProperty("timestamp_ns").GetInt64());
            Assert.True(ok, $"{curr.GetProperty("description")} after {prev.GetProperty("description")}");
        }

        // Equal timestamp must fail.
        var v = vectors[0];
        var ts = v.GetProperty("timestamp_ns").GetInt64();
        Assert.False(Verifier.VerifyTreeHeadSequential(
            (ulong)v.GetProperty("tree_size").GetInt64(),
            H(v.GetProperty("root_hash_hex").GetString()!),
            ts,
            H(v.GetProperty("signature_hex").GetString()!),
            H(v.GetProperty("public_key_hex").GetString()!),
            ts), "equal timestamp must fail");
    }

    // ── Negative cases ────────────────────────────────────────────────────────

    [Fact]
    public void NegativeCases()
    {
        foreach (var kv in Load("negative-cases.json").EnumerateObject())
        {
            var v = kv.Value;
            var expected = v.GetProperty("expected_result").GetBoolean();
            if (kv.Name.StartsWith("inclusion_"))
            {
                var got = Verifier.VerifyInclusion(
                    H(v.GetProperty("leaf_hash_hex").GetString()!),
                    (ulong)v.GetProperty("leaf_index").GetInt64(),
                    (ulong)v.GetProperty("tree_size").GetInt64(),
                    Proofs(v.GetProperty("proof_hex")),
                    H(v.GetProperty("root_hex").GetString()!));
                Assert.Equal(expected, got);
            }
            else if (kv.Name.StartsWith("consistency_"))
            {
                var got = Verifier.VerifyConsistency(
                    (ulong)v.GetProperty("old_size").GetInt64(),
                    (ulong)v.GetProperty("new_size").GetInt64(),
                    Proofs(v.GetProperty("proof_hex")),
                    H(v.GetProperty("old_root_hex").GetString()!),
                    H(v.GetProperty("new_root_hex").GetString()!));
                Assert.Equal(expected, got);
            }
            else if (kv.Name.StartsWith("tree_head_sequential_"))
            {
                var got = Verifier.VerifyTreeHeadSequential(
                    (ulong)v.GetProperty("tree_size").GetInt64(),
                    H(v.GetProperty("root_hash_hex").GetString()!),
                    v.GetProperty("timestamp_ns").GetInt64(),
                    H(v.GetProperty("signature_hex").GetString()!),
                    H(v.GetProperty("public_key_hex").GetString()!),
                    v.GetProperty("previous_timestamp_ns").GetInt64());
                Assert.Equal(expected, got);
            }
        }
    }

    private static string ToHex(byte[] bytes) =>
        BitConverter.ToString(bytes).Replace("-", "").ToLower();
}

File: sdks/rust/tests/verifier_tests.rs
use settled_sdk::verifier::{
    leaf_hash, node_hash, verify_consistency, verify_inclusion, verify_tree_head,
    verify_tree_head_sequential,
};
use std::path::PathBuf;

fn vectors_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../test-vectors")
}

fn load_json(name: &str) -> Vec<u8> {
    std::fs::read(vectors_dir().join(name)).unwrap_or_else(|e| panic!("cannot read {name}: {e}"))
}

fn h(s: &str) -> Vec<u8> {
    hex::decode(s).unwrap_or_else(|_| panic!("invalid hex: {s}"))
}

fn b32(s: &str) -> [u8; 32] {
    h(s).try_into().expect("expected 32 bytes")
}

fn proof(hexes: &[String]) -> Vec<[u8; 32]> {
    hexes.iter().map(|s| b32(s)).collect()
}

// ── Leaf hashes ───────────────────────────────────────────────────────────────

#[test]
fn test_leaf_hash() {
    #[derive(serde::Deserialize)]
    struct Vector {
        description: String,
        input_hex: String,
        hash_hex: String,
    }
    let vectors: Vec<Vector> = serde_json::from_slice(&load_json("leaf-hashes.json")).unwrap();
    for v in vectors {
        let got = leaf_hash(&h(&v.input_hex));
        assert_eq!(hex::encode(got), v.hash_hex, "{}", v.description);
    }
}

// ── Node hashes ───────────────────────────────────────────────────────────────

#[test]
fn test_node_hash() {
    #[derive(serde::Deserialize)]
    struct Vector {
        description: String,
        left_hex: String,
        right_hex: String,
        hash_hex: Option<String>,
        swapped_hash_hex: Option<String>,
    }
    let vectors: Vec<Vector> = serde_json::from_slice(&load_json("node-hashes.json")).unwrap();
    for v in vectors {
        let left = b32(&v.left_hex);
        let right = b32(&v.right_hex);
        if let Some(expected) = v.hash_hex {
            let got = node_hash(&left, &right);
            assert_eq!(hex::encode(got), expected, "{}", v.description);
        }
        if let Some(expected_swapped) = v.swapped_hash_hex {
            let ab = node_hash(&left, &right);
            let ba = node_hash(&right, &left);
            assert_ne!(ab, ba, "{} must not be commutative", v.description);
            assert_eq!(
                hex::encode(ba),
                expected_swapped,
                "{} swapped",
                v.description
            );
        }
    }
}

// ── Inclusion proofs ──────────────────────────────────────────────────────────

#[test]
fn test_verify_inclusion() {
    #[derive(serde::Deserialize)]
    struct Vector {
        tree_size: u64,
        leaf_index: u64,
        leaf_hash_hex: String,
        proof_hex: Vec<String>,
        root_hex: String,
    }
    let vectors: Vec<Vector> = serde_json::from_slice(&load_json("inclusion-proofs.json")).unwrap();
    for v in vectors {
        let ok = verify_inclusion(
            b32(&v.leaf_hash_hex),
            v.leaf_index,
            v.tree_size,
            &proof(&v.proof_hex),
            b32(&v.root_hex),
        );
        assert!(ok, "size={} idx={}", v.tree_size, v.leaf_index);
    }
}

// ── Consistency proofs ────────────────────────────────────────────────────────

#[test]
fn test_verify_consistency() {
    #[derive(serde::Deserialize)]
    struct Vector {
        old_size: u64,
        new_size: u64,
        old_root_hex: String,
        new_root_hex: String,
        proof_hex: Vec<String>,
    }
    let vectors: Vec<Vector> =
        serde_json::from_slice(&load_json("consistency-proofs.json")).unwrap();
    for v in vectors {
        let ok = verify_consistency(
            v.old_size,
            v.new_size,
            &proof(&v.proof_hex),
            b32(&v.old_root_hex),
            b32(&v.new_root_hex),
        );
        assert!(ok, "old={} new={}", v.old_size, v.new_size);
    }
}

// ── Signed Tree Heads ─────────────────────────────────────────────────────────

#[test]
fn test_verify_tree_head() {
    #[derive(serde::Deserialize)]
    struct Vector {
        description: String,
        tree_size: u64,
        root_hash_hex: String,
        timestamp_ns: i64,
        signature_hex: String,
        public_key_hex: String,
    }
    let vectors: Vec<Vector> =
        serde_json::from_slice(&load_json("signed-tree-heads.json")).unwrap();
    for v in vectors {
        assert!(
            verify_tree_head(
                v.tree_size,
                b32(&v.root_hash_hex),
                v.timestamp_ns,
                &h(&v.signature_hex),
                &h(&v.public_key_hex)
            ),
            "{}",
            v.description,
        );

        assert!(
            !verify_tree_head(
                v.tree_size + 1,
                b32(&v.root_hash_hex),
                v.timestamp_ns,
                &h(&v.signature_hex),
                &h(&v.public_key_hex)
            ),
            "{} tampered tree_size should fail",
            v.description,
        );

        let mut root = b32(&v.root_hash_hex);
        root[0] ^= 0xFF;
        assert!(
            !verify_tree_head(
                v.tree_size,
                root,
                v.timestamp_ns,
                &h(&v.signature_hex),
                &h(&v.public_key_hex)
            ),
            "{} tampered root should fail",
            v.description,
        );
    }
}

// ── Sequential STH verification ───────────────────────────────────────────────

#[test]
fn test_verify_tree_head_sequential() {
    #[derive(serde::Deserialize)]
    struct Vector {
        description: String,
        tree_size: u64,
        root_hash_hex: String,
        timestamp_ns: i64,
        signature_hex: String,
        public_key_hex: String,
    }
    let vectors: Vec<Vector> =
        serde_json::from_slice(&load_json("signed-tree-heads.json")).unwrap();

    // Consecutive pairs must pass sequential verification (timestamps are strictly increasing).
    for pair in vectors.windows(2) {
        let prev = &pair[0];
        let curr = &pair[1];
        assert!(
            verify_tree_head_sequential(
                curr.tree_size,
                b32(&curr.root_hash_hex),
                curr.timestamp_ns,
                &h(&curr.signature_hex),
                &h(&curr.public_key_hex),
                prev.timestamp_ns,
            ),
            "{} after {}",
            curr.description,
            prev.description,
        );
    }

    // Same STH with itself as previous must fail (equal timestamp).
    let v = &vectors[0];
    assert!(
        !verify_tree_head_sequential(
            v.tree_size,
            b32(&v.root_hash_hex),
            v.timestamp_ns,
            &h(&v.signature_hex),
            &h(&v.public_key_hex),
            v.timestamp_ns,
        ),
        "equal timestamp must fail",
    );
}

// ── Negative cases ────────────────────────────────────────────────────────────

#[test]
fn test_negative_cases() {
    use std::collections::HashMap;

    let cases: HashMap<String, serde_json::Value> =
        serde_json::from_slice(&load_json("negative-cases.json")).unwrap();

    for (name, v) in &cases {
        let expected = v["expected_result"].as_bool().unwrap();

        if name.starts_with("inclusion_") {
            let got = verify_inclusion(
                b32(v["leaf_hash_hex"].as_str().unwrap()),
                v["leaf_index"].as_u64().unwrap(),
                v["tree_size"].as_u64().unwrap(),
                &v["proof_hex"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .map(|s| b32(s.as_str().unwrap()))
                    .collect::<Vec<_>>(),
                b32(v["root_hex"].as_str().unwrap()),
            );
            assert_eq!(got, expected, "{name}");
        } else if name.starts_with("consistency_") {
            let got = verify_consistency(
                v["old_size"].as_u64().unwrap(),
                v["new_size"].as_u64().unwrap(),
                &v["proof_hex"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .map(|s| b32(s.as_str().unwrap()))
                    .collect::<Vec<_>>(),
                b32(v["old_root_hex"].as_str().unwrap()),
                b32(v["new_root_hex"].as_str().unwrap()),
            );
            assert_eq!(got, expected, "{name}");
        } else if name.starts_with("tree_head_sequential_") {
            let got = verify_tree_head_sequential(
                v["tree_size"].as_u64().unwrap(),
                b32(v["root_hash_hex"].as_str().unwrap()),
                v["timestamp_ns"].as_i64().unwrap(),
                &h(v["signature_hex"].as_str().unwrap()),
                &h(v["public_key_hex"].as_str().unwrap()),
                v["previous_timestamp_ns"].as_i64().unwrap(),
            );
            assert_eq!(got, expected, "{name}");
        }
    }
}
